src/Lexer.php

```diff
@@ -451,8 +446,7 @@ private function stripcslashes(string $str, string $quoteType): string
                 }
                 $result .= chr(octdec($octal));
             } else {
-                trigger_deprecation('twig/twig', '3.12', sprintf('Character "%s" at position %d does not need to be escaped anymore.', $nextChar, $i + 1));
-                $result .= $nextChar;
+                $result .= '\\' . $nextChar;
```

Comment:
I think there is an issue here.

I've tested with the following template:

```twig
{{ 'Foo\Bar' }}
{{ 'Foo\\Bar' }}
{{ 'Foo\\\Bar' }}
{{ 'Foo\\\\Bar' }}
```

On Twig 3.x, it results in

```twig
FooBar
Foo\Bar
Foo\Bar
Foo\\Bar
```

But with this PR, I get:

```twig
Foo\Bar
Foo\Bar
Foo\\Bar
Foo\\Bar
```

To fix it:

```suggestion
                $result .= $nextChar;
```

Reply:
I think this is correct.

Escaping B is unnecessary. In v3 a deprecation is reported. In v4 it prints `\`.
```twig
{{ 'Foo\Bar' }}
v3: FooBar
v4: Foo\Bar
```

Escaping a backslash is ignored in v3 and still ignored in v4:
```twig
{{ 'Foo\\Bar' }}
v3: Foo\Bar
v4: Foo\Bar
```

Escaping a backslash is ignored in v3 and still ignored in v4. 
Escaping B is unnecessary. In v3 a deprecation is reported. In v4 it prints `\`.
```twig
{{ 'Foo\\\Bar' }}
v3: Foo\Bar
v4: Foo\\Bar
```

Escaping a backslash is ignored in v3 and still ignored in v4:
```twig
{{ 'Foo\\\\Bar' }}
v3: Foo\\Bar
v4: Foo\\Bar
```

Reply:
Added an integration test to better show it:
https://github.com/twigphp/Twig/pull/4199/commits/152990db5f662e3edb1bdb2248ee5436f68dc6aa

Reply:
We still need to allow escaping a backslash, otherwise writing a backslash followed by a `n` is impossible (as `\n` is a valid escape sequence.

The new behavior matches the behavior of PHP strings: escaping of backslashes is _optional_ when non-ambiguous. But if you want to write a backslash when the next char can be an escaped char (which includes the case when you want to write a backslash followed by another one), it needs to be escaped.